---
title: "Courses involving phyloseq"
date: "`r date()`"
output: 
  html_document: 
    toc: yes
    toc_depth: 2
    toc_float: true
---

<link rel="stylesheet" href="site_libs/academicons-1.7.0/css/academicons.css"/>

&nbsp;

---


There are a number of courses now including or featuring phyloseq. I will try and link here to the ones I know about.

---

# 2018

&nbsp;

## [ASM Microbe Workshop](https://www.asm.org/index.php/program-microbe-2018/courses-and-workshops)

- [006-WS: Microbe Analysis Using R](http://www.abstractsonline.com/pp8/#!/4623/session/1159)

Session Type: Full-Day

Time: 8:00 a.m. – 4:00 p.m. 

The workshop will cover the use of a variety of R packages useful for the analysis of amplicon sequencing data (esp. 16S rRNA gene targets).
These analysis are built around phyloseq data objects and extend on basic concepts in microbiome analysis 
such as alpha and beta diversity with inclusion of advanced material on differential abundance testing, 
feature selection and time-series analysis.

[REGISTER HERE](https://registration.experientevent.com/showasm181/Flow/ATT?&_ga=2.234769822.1604892224.1520870826-1543952640.1510599824#/registrant//Dashboard/)

&nbsp;

&nbsp;


## [SISMID 2018](http://www.biostat.washington.edu/suminst/sismid)

- [SISMID Module 8: Microbiome Data Analysis](http://www.biostat.washington.edu/suminst/sismid2018/modules/MD1808)

- Session 3: Mon Jul 16 to Wed Jul 18

Instructor(s):

Alexander Alekseyenko and Paul McMurdie
Module dates/times: 
- Monday, July 16; 8:30 a.m. - 5 p.m. 
- Tuesday, July 17, 8:30 a.m.-5 p.m. 
- Wednesday, July 18, 8:30 a.m.-Noon

Prerequisites: Programming will be done in R and fluency at the level of the module on Introduction to R, though not necessarily from taking that module, will be expected. This module assumes knowledge of the material in Module 1: Probability and Statistical Inference, though not necessarily from taking that module.

This course is concerned with multivariate statistical analysis of microbiome data. We will briefly cover foundational concepts in microbial ecology, molecular biology, bioinformatics, and DNA sequencing.

The main focus of the course will be on developing an understanding of multivariate analysis of microbiome data. Practical skills to be developed in this course include managing high-dimensional and structured data in metagenomics, visualization and representation of high-dimensional data, normalization, filtering, and mixture-model noise modeling of count data, as well as clustering and predictive model building.

&nbsp;

&nbsp;


## [STAMPS 2018](http://www.mbl.edu/education/courses/stamps/)

- Strategies and Techniques for Analyzing Microbial Population Structures (STAMPS)
Course Information

- Course Dates: July 29 – August 8, 2018
- Registration Deadline: April 6, 2018 | [Apply here](https://ws2.mbl.edu/studentapp/studentapp.asp?courseID=STAMPS)

Directors: Mihai Pop, University of Maryland; and Tracy Teal, Michigan State University

&nbsp;

&nbsp;

&nbsp;

&nbsp;

---

# 2017

&nbsp;

## Workshop on Genomics, 2017

- [Workshop on Genomics, Český Krumlov, Czech Republic (2017 January)](http://evomics.org/workshops/workshop-on-genomics/)

The Workshop on Genomics was developed in response to the increasing demand for training on how to effectively analyze and manage data generated by modern sequencing technologies. The Workshop curriculum includes extensive coverage of fundamental techniques required of all studies utilizing modern sequence data. This includes management and modification of modern sequence data files, using the Unix command-line and editors to manipulate sequence data files, assembly (both mapping and de novo) and best practices for sequence data generation. The second-half of the Workshop is devoted to specific sub-disciplines within the genomic sciences. While these topics may change from Workshop to Workshop, efforts are made to devote significant amounts of time to techniques being used widely in the field (e.g. transcriptomics, genome assembly, metagenomics).

A majority of the schedule is dedicated to hands-on learning activities designed by faculty and the workshop team. This interactive experience provides Workshop participants with the practical experience required to meet the challenges presented by modern genomic sciences.


## [SISMID, 2017](http://www.biostat.washington.edu/suminst/sismid2017/modules/MD1706)

- [SISMID Module 6: Microbiome Data Analysis](http://www.biostat.washington.edu/suminst/sismid2017/modules/MD1706)

Week 1, Session 2

`Wed Jul 12 1:30 PM - Fri Jul 14 5:00 PM`

Instructor(s):
Alexander Alekseyenko and Paul McMurdie

Programming will be done in R and fluency at the level of *Module 4: Introduction to R* will be expected.

**Pre-requisites**: Knowledge of *Module 4: Introduction to R*.

This course is concerned with multivariate statistical analysis of microbiome data. We will briefly cover foundational concepts in microbial ecology, molecular biology, bioinformatics, and DNA sequencing.

The main focus of the course will be on developing an understanding of multivariate analysis of microbiome data. Practical skills to be developed in this course include managing high-dimensional and structured data in metagenomics, visualization and representation of high-dimensional data, normalization, filtering, and mixture-model noise modeling of count data, as well as clustering and predictive model building.

&nbsp;

## [STAMPS 2017](https://stamps.mbl.edu/index.php/Main_Page)

I understand this course is going on again this year, but the web page was not updated at the time I wrote this.

My unofficial information on scheduling is:

Evening of Sunday July 31 -- morning on Wednesday Aug 9.

&nbsp;


&nbsp;

&nbsp;

---

# 2016

&nbsp;

## SISMID Workshop, 2016

- [Summer Institute in Statistics, Modeling in Infectious Diseases (SISMID): (2016 July 25th to July 27th) Module 14, Introduction to Metagenomic Data Analysis](http://www.biostat.washington.edu/suminst/sismid2016/modules/MD1614)

Week 3, Session 5

Monday 8:30 AM - Wednesday 12:00 PM: Mon Jul 25 to Wed Jul 27

Instructor(s):
Alexander Alekseyenko and Paul McMurdie

This course is concerned with multivariate statistical analysis of microbiome data. We will briefly cover foundational concepts in microbial ecology, molecular biology, bioinformatics, and DNA sequencing. The main focus of the course will be on developing an understanding of multivariate analysis of microbiome data. Practical skills to be developed in this course include managing high-dimensional and structured data in metagenomics, visualization and representation of high-dimensional data, normalization, filtering, and mixture-model noise modeling of count data, as well as clustering and predictive model building. Programming will be done in R and fluency at the level of ‘SISMID/SISG Module 4: Introduction to R’ will be expected. Pre-requisites: knowledge of Module 1, Probability and Statistical Inference.


&nbsp;


## STAMPS Workshop, 2016

- Strategies and Techniques for Analyzing Microbial Population Structures (STAMPS)

Course Dates: August 3 – August 13, 2016

Deadline: April 8, 2016 (Expired!)

[Course Website](https://stamps.mbl.edu/index.php/Main_Page)

[Course Schedule](https://stamps.mbl.edu/index.php/Schedule)


&nbsp;


## Workshop on Genomics, 2016

- [Workshop on Genomics, Cesky Krumlov, Czech Republic (2016 Jan 10 - 23)](http://evomics.org/workshops/workshop-on-genomics/2016-workshop-on-genomics-cesky-krumlov/)



&nbsp;

&nbsp;

&nbsp;

---
# 2015

## Workshop on Metagenomics (Harvard)

- [Harvard University Workshop on Metagenomics (2015 Sept 14 - 18)](http://evomics.org/workshops/harvard-university-workshops/2015-harvard-university-workshop-on-metagenomics/)

A week-long workshop focused exclusively on the analysis of metagenomic data.  This intensive, hands-on course will feature morning and afternoon sessions taught by expert faculty and technical assistants from the Broad, Ragon Institute, Harvard School of Public Health, Stanford University and Washington University, St. Louis.  All participants will acquire fundamental computational skills, appropriate study design, data interpretation, and overall literacy in bioinformatics.

When: 2015 14-18, September : Morning Sessions 9:00 – 12:00 | Afternoon Sessions 2:00 – 5:00

Where: Joseph B. Martin Conference Center Harvard Medical School 77 Avenue Louis Pasteur Boston, MA 02115.


&nbsp;


## SISMID Workshop, 2015

- [Summer Institute in Statistics, Modeling in Infectious Diseases (SISMID): Metagenomics Module](http://www.biostat.washington.edu/suminst/sismid/modules)


&nbsp;

## STAMPS Workshop, 2015

- [Strategies and Techniques for Analyzing Microbial Population Structure (STAMPS)](https://stamps.mbl.edu/index.php/Main_Page)

Course Date: August 5 - August 15, 2015

Twitter hash: #STAMPS15
