<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Courses involving phyloseq</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link href="site_libs/ionicons-2.0.1/css/ionicons.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<div class="container-fluid main-container">

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">phyloseq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="install.html">
    <span class="ion ion-android-archive"></span>
     
    Installation
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-newspaper-o"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0061217">phyloseq in Bioconductor</a>
    </li>
    <li>
      <a href="http://bioinformatics.oxfordjournals.org/content/early/2014/09/26/bioinformatics.btu616">Shiny-phyloseq</a>
    </li>
    <li>
      <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003531">Rarefying Inadmissible</a>
    </li>
    <li>
      <a href="http://f1000research.com/articles/5-1492/v1">Full Stack Workflow</a>
    </li>
    <li>
      <a href="https://www.nature.com/articles/nmeth.3869">DADA2</a>
    </li>
    <li>
      <a href="https://www.nature.com/articles/ismej2017119">Amplicon Sequence Variants &gt; OTUs</a>
    </li>
  </ul>
</li>
<li>
  <a href="courses.html">
    <span class="fa fa-graduation-cap"></span>
     
    Courses
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-university"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Basics</li>
    <li>
      <a href="import-data.html">Data Import</a>
    </li>
    <li>
      <a href="Example-Data.html">Example Data</a>
    </li>
    <li>
      <a href="merge.html">Merge</a>
    </li>
    <li>
      <a href="preprocess.html">Preprocess</a>
    </li>
    <li>
      <a href="distance.html">Distances</a>
    </li>
    <li>
      <a href="gap-statistic.html">Gap Statistic</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Workflows</li>
    <li>
      <a href="http://benjjneb.github.io/dada2/tutorial.html">DADA2</a>
    </li>
    <li>
      <a href="http://f1000research.com/articles/5-1492/v1">Full Stack</a>
    </li>
    <li>
      <a href="http://joey711.github.io/phyloseq-demo/HMP_import_example.html">HMP-1</a>
    </li>
    <li>
      <a href="http://joey711.github.io/phyloseq-demo/Restroom-Biogeography.html">Restroom Biogeography</a>
    </li>
    <li>
      <a href="http://joey711.github.io/phyloseq-demo/unifrac.html">UniFrac</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Graphics</li>
    <li>
      <a href="plot_ordination-examples.html">Plot Ordination</a>
    </li>
    <li>
      <a href="plot_richness-examples.html">Plot Alpha Diversity</a>
    </li>
    <li>
      <a href="plot_tree-examples.html">Plot Tree</a>
    </li>
    <li>
      <a href="plot_bar-examples.html">
        <span class="fa fa-bar-chart"></span>
         
        Plot Bars
      </a>
    </li>
    <li>
      <a href="plot_heatmap-examples.html">
        <span class="fa fa-qrcode"></span>
         
        Plot Heatmap
      </a>
    </li>
    <li>
      <a href="plot_network-examples.html">
        <span class="ion ion-network"></span>
         
        Plot Network
      </a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Differential Abundance</li>
    <li>
      <a href="http://joey711.github.io/phyloseq-extensions/DESeq2.html">DESeq2 (Recommended)</a>
    </li>
    <li>
      <a href="http://joey711.github.io/phyloseq-extensions/edgeR.html">edgeR</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/joey711/phyloseq/issues">
    <span class="fa fa-medkit"></span>
     
    Issues
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/phyloseq/inst/doc/phyloseq-FAQ.html">
    <span class="fa fa-question"></span>
     
    FAQ
  </a>
</li>
<li>
  <a href="https://github.com/joey711/phyloseq">
    <span class="fa fa-github-alt"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-twitter"></span>
     
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://twitter.com/joey711">
        <span class="fa fa-twitter"></span>
         
      </a>
    </li>
    <li>
      <a href="https://www.linkedin.com/in/joey711">
        <span class="fa fa-linkedin-square"></span>
         
      </a>
    </li>
    <li>
      <a href="http://stackoverflow.com/users/935950/paul-mcmurdie">
        <span class="fa fa-stack-overflow"></span>
         
      </a>
    </li>
    <li>
      <a href="https://scholar.google.com/citations?user=nNQ1LtMAAAAJ">
        <span class="ai ai-google-scholar-square"></span>
         
      </a>
    </li>
    <li>
      <a href="https://www.researchgate.net/profile/Paul_Mcmurdie">
        <span class="ai ai-researchgate-square"></span>
         
      </a>
    </li>
    <li>
      <a href="https://github.com/joey711/phyloseq">
        <span class="fa fa-github-alt"></span>
         
      </a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Courses involving phyloseq</h1>
<h4 class="date"><em>Sun Feb 24 14:54:33 2019</em></h4>

</div>


<p><link rel="stylesheet" href="site_libs/academicons-1.7.0/css/academicons.css"/></p>
<p> </p>
<hr />
<p>There are a number of courses now including or featuring phyloseq. I will try and link here to the ones I know about.</p>
<hr />
<div id="section" class="section level1">
<h1>2019</h1>
<p> </p>
<div id="sismid-2019" class="section level2">
<h2><a href="https://www.biostat.washington.edu/suminst/sismid">SISMID 2019</a></h2>
<ul>
<li><p><a href="https://www.biostat.washington.edu/suminst/sismid2019/modules/MD1908">SISMID Module 8: Microbiome Data Analysis</a></p></li>
<li><p>Session 3:</p></li>
</ul>
<blockquote>
<p>Mon Jul 15 to Wed Jul 17</p>
</blockquote>
<p>Instructor(s):</p>
<p>Alexander Alekseyenko and Paul McMurdie</p>
<p>Module dates/times:</p>
<ul>
<li>Monday, July 15; 8:30 a.m. - 5 p.m.</li>
<li>Tuesday, July 16, 8:30 a.m.-5 p.m.</li>
<li>Wednesday, July 17, 8:30 a.m.-Noon</li>
</ul>
<p>Prerequisites:</p>
<p>Programming will be done in R and fluency at the level of the module on Introduction to R, though not necessarily from taking that module, will be expected. This module assumes knowledge of the material in Module 1: Probability and Statistical Inference, though not necessarily from taking that module.</p>
<p>This course is concerned with multivariate statistical analysis of microbiome data. We will briefly cover foundational concepts in microbial ecology, molecular biology, bioinformatics, and DNA sequencing.</p>
<p>The main focus of the course will be on developing an understanding of multivariate analysis of microbiome data. Practical skills to be developed in this course include managing high-dimensional and structured data in metagenomics, visualization and representation of high-dimensional data, normalization, filtering, and mixture-model noise modeling of count data, as well as clustering and predictive model building.</p>
<p> </p>
<p> </p>
</div>
<div id="asm-microbe-workshop" class="section level2">
<h2><a href="https://www.abstractsonline.com/pp8/#!/7859/session/357">ASM Microbe Workshop</a></h2>
<ul>
<li><a href="https://www.abstractsonline.com/pp8/#!/7859/session/357">CS-003 - Microbial Ecology using R and Bioconductor: From Raw Reads to Statistical Testing</a></li>
</ul>
<p>Session Type: Full-Day</p>
<blockquote>
<p>Mon 2019 June 20</p>
</blockquote>
<p>Time: 8:00 a.m. – 4:00 p.m.</p>
<blockquote>
<p>Current microbial ecology studies encounter the challenges of all modern biological data: high levels of heterogeneity, sub-optimal experimental design, and high dimensionality. In this course, instructors will introduce approaches to deal with these challenges with two commonly used computational tools: R and Bioconductor. Participants will receive an introduction to R and packages used for data manipulation and visualization (e.g., ggplot2). Other sessions and interactive labs will focus on the common practices involved in the processing amplicon sequence data to make ecological inference.</p>
</blockquote>
<p>*Please Note: Attendees are required to bring their own laptops to participate in this course For R beginners. The Faculty of this Course recommend looking through a few of these links to get you a basic proficiency in order to maximize learning during the course:</p>
<ul>
<li><a href="https://www.edx.org/course/statistics-r-harvardx-ph525-1x-1" class="uri">https://www.edx.org/course/statistics-r-harvardx-ph525-1x-1</a></li>
<li><a href="https://www.edx.org/course/r-basics" class="uri">https://www.edx.org/course/r-basics</a></li>
<li><a href="https://www.rstudio.com/online-learning/#r-programming" class="uri">https://www.rstudio.com/online-learning/#r-programming</a></li>
<li><a href="https://support.rstudio.com/hc/en-us/articles/201141096-Getting-Started-with-R" class="uri">https://support.rstudio.com/hc/en-us/articles/201141096-Getting-Started-with-R</a></li>
<li><a href="https://monashdatafluency.github.io/r-intro-2/" class="uri">https://monashdatafluency.github.io/r-intro-2/</a></li>
<li><a href="https://www.kdnuggets.com/2016/03/datacamp-r-learning-path-7-steps.html" class="uri">https://www.kdnuggets.com/2016/03/datacamp-r-learning-path-7-steps.html</a></li>
</ul>
<p>The workshop will cover the use of a variety of R packages useful for the analysis of amplicon sequencing data (esp. 16S rRNA gene targets). These analyses will cover details of sequence data processing using <a href="https://benjjneb.github.io/dada2/index.html">DADA2</a>, creating a phyloseq data object, and extended concepts in microbiome analysis.</p>
<p> </p>
<p> </p>
<hr />
</div>
</div>
<div id="section-1" class="section level1">
<h1>2018</h1>
<p> </p>
<div id="asm-microbe-workshop-1" class="section level2">
<h2><a href="https://www.asm.org/index.php/program-microbe-2018/courses-and-workshops">ASM Microbe Workshop</a></h2>
<ul>
<li><a href="http://www.abstractsonline.com/pp8/#!/4623/session/1159">006-WS: Microbe Analysis Using R</a></li>
</ul>
<p>Session Type: Full-Day</p>
<p>Time: 8:00 a.m. – 4:00 p.m.</p>
<p>The workshop will cover the use of a variety of R packages useful for the analysis of amplicon sequencing data (esp. 16S rRNA gene targets). These analysis are built around phyloseq data objects and extend on basic concepts in microbiome analysis such as alpha and beta diversity with inclusion of advanced material on differential abundance testing, feature selection and time-series analysis.</p>
<p><a href="https://registration.experientevent.com/showasm181/Flow/ATT?&amp;_ga=2.234769822.1604892224.1520870826-1543952640.1510599824#/registrant//Dashboard/">REGISTER HERE</a></p>
<p> </p>
<p> </p>
</div>
<div id="sismid-2018" class="section level2">
<h2><a href="http://www.biostat.washington.edu/suminst/sismid">SISMID 2018</a></h2>
<ul>
<li><p><a href="http://www.biostat.washington.edu/suminst/sismid2018/modules/MD1808">SISMID Module 8: Microbiome Data Analysis</a></p></li>
<li><p>Session 3: Mon Jul 16 to Wed Jul 18</p></li>
</ul>
<p>Instructor(s):</p>
<p>Alexander Alekseyenko and Paul McMurdie Module dates/times: - Monday, July 16; 8:30 a.m. - 5 p.m. - Tuesday, July 17, 8:30 a.m.-5 p.m. - Wednesday, July 18, 8:30 a.m.-Noon</p>
<p>Prerequisites: Programming will be done in R and fluency at the level of the module on Introduction to R, though not necessarily from taking that module, will be expected. This module assumes knowledge of the material in Module 1: Probability and Statistical Inference, though not necessarily from taking that module.</p>
<p>This course is concerned with multivariate statistical analysis of microbiome data. We will briefly cover foundational concepts in microbial ecology, molecular biology, bioinformatics, and DNA sequencing.</p>
<p>The main focus of the course will be on developing an understanding of multivariate analysis of microbiome data. Practical skills to be developed in this course include managing high-dimensional and structured data in metagenomics, visualization and representation of high-dimensional data, normalization, filtering, and mixture-model noise modeling of count data, as well as clustering and predictive model building.</p>
<p> </p>
<p> </p>
</div>
<div id="stamps-2018" class="section level2">
<h2><a href="http://www.mbl.edu/education/courses/stamps/">STAMPS 2018</a></h2>
<ul>
<li><p>Strategies and Techniques for Analyzing Microbial Population Structures (STAMPS) Course Information</p></li>
<li>Course Dates: July 29 – August 8, 2018</li>
<li><p>Registration Deadline: April 6, 2018 | <a href="https://ws2.mbl.edu/studentapp/studentapp.asp?courseID=STAMPS">Apply here</a></p></li>
</ul>
<p>Directors: Mihai Pop, University of Maryland; and Tracy Teal, Michigan State University</p>
<p> </p>
<p> </p>
<p> </p>
<p> </p>
<hr />
</div>
</div>
<div id="section-2" class="section level1">
<h1>2017</h1>
<p> </p>
<div id="workshop-on-genomics-2017" class="section level2">
<h2>Workshop on Genomics, 2017</h2>
<ul>
<li><a href="http://evomics.org/workshops/workshop-on-genomics/">Workshop on Genomics, Český Krumlov, Czech Republic (2017 January)</a></li>
</ul>
<p>The Workshop on Genomics was developed in response to the increasing demand for training on how to effectively analyze and manage data generated by modern sequencing technologies. The Workshop curriculum includes extensive coverage of fundamental techniques required of all studies utilizing modern sequence data. This includes management and modification of modern sequence data files, using the Unix command-line and editors to manipulate sequence data files, assembly (both mapping and de novo) and best practices for sequence data generation. The second-half of the Workshop is devoted to specific sub-disciplines within the genomic sciences. While these topics may change from Workshop to Workshop, efforts are made to devote significant amounts of time to techniques being used widely in the field (e.g. transcriptomics, genome assembly, metagenomics).</p>
<p>A majority of the schedule is dedicated to hands-on learning activities designed by faculty and the workshop team. This interactive experience provides Workshop participants with the practical experience required to meet the challenges presented by modern genomic sciences.</p>
</div>
<div id="sismid-2017" class="section level2">
<h2><a href="http://www.biostat.washington.edu/suminst/sismid2017/modules/MD1706">SISMID, 2017</a></h2>
<ul>
<li><a href="http://www.biostat.washington.edu/suminst/sismid2017/modules/MD1706">SISMID Module 6: Microbiome Data Analysis</a></li>
</ul>
<p>Week 1, Session 2</p>
<p><code>Wed Jul 12 1:30 PM - Fri Jul 14 5:00 PM</code></p>
<p>Instructor(s): Alexander Alekseyenko and Paul McMurdie</p>
<p>Programming will be done in R and fluency at the level of <em>Module 4: Introduction to R</em> will be expected.</p>
<p><strong>Pre-requisites</strong>: Knowledge of <em>Module 4: Introduction to R</em>.</p>
<p>This course is concerned with multivariate statistical analysis of microbiome data. We will briefly cover foundational concepts in microbial ecology, molecular biology, bioinformatics, and DNA sequencing.</p>
<p>The main focus of the course will be on developing an understanding of multivariate analysis of microbiome data. Practical skills to be developed in this course include managing high-dimensional and structured data in metagenomics, visualization and representation of high-dimensional data, normalization, filtering, and mixture-model noise modeling of count data, as well as clustering and predictive model building.</p>
<p> </p>
</div>
<div id="stamps-2017" class="section level2">
<h2><a href="https://stamps.mbl.edu/index.php/Main_Page">STAMPS 2017</a></h2>
<p>I understand this course is going on again this year, but the web page was not updated at the time I wrote this.</p>
<p>My unofficial information on scheduling is:</p>
<p>Evening of Sunday July 31 – morning on Wednesday Aug 9.</p>
<p> </p>
<p> </p>
<p> </p>
<hr />
</div>
</div>
<div id="section-3" class="section level1">
<h1>2016</h1>
<p> </p>
<div id="sismid-workshop-2016" class="section level2">
<h2>SISMID Workshop, 2016</h2>
<ul>
<li><a href="http://www.biostat.washington.edu/suminst/sismid2016/modules/MD1614">Summer Institute in Statistics, Modeling in Infectious Diseases (SISMID): (2016 July 25th to July 27th) Module 14, Introduction to Metagenomic Data Analysis</a></li>
</ul>
<p>Week 3, Session 5</p>
<p>Monday 8:30 AM - Wednesday 12:00 PM: Mon Jul 25 to Wed Jul 27</p>
<p>Instructor(s): Alexander Alekseyenko and Paul McMurdie</p>
<p>This course is concerned with multivariate statistical analysis of microbiome data. We will briefly cover foundational concepts in microbial ecology, molecular biology, bioinformatics, and DNA sequencing. The main focus of the course will be on developing an understanding of multivariate analysis of microbiome data. Practical skills to be developed in this course include managing high-dimensional and structured data in metagenomics, visualization and representation of high-dimensional data, normalization, filtering, and mixture-model noise modeling of count data, as well as clustering and predictive model building. Programming will be done in R and fluency at the level of ‘SISMID/SISG Module 4: Introduction to R’ will be expected. Pre-requisites: knowledge of Module 1, Probability and Statistical Inference.</p>
<p> </p>
</div>
<div id="stamps-workshop-2016" class="section level2">
<h2>STAMPS Workshop, 2016</h2>
<ul>
<li>Strategies and Techniques for Analyzing Microbial Population Structures (STAMPS)</li>
</ul>
<p>Course Dates: August 3 – August 13, 2016</p>
<p>Deadline: April 8, 2016 (Expired!)</p>
<p><a href="https://stamps.mbl.edu/index.php/Main_Page">Course Website</a></p>
<p><a href="https://stamps.mbl.edu/index.php/Schedule">Course Schedule</a></p>
<p> </p>
</div>
<div id="workshop-on-genomics-2016" class="section level2">
<h2>Workshop on Genomics, 2016</h2>
<ul>
<li><a href="http://evomics.org/workshops/workshop-on-genomics/2016-workshop-on-genomics-cesky-krumlov/">Workshop on Genomics, Cesky Krumlov, Czech Republic (2016 Jan 10 - 23)</a></li>
</ul>
<p> </p>
<p> </p>
<p> </p>
<hr />
</div>
</div>
<div id="section-4" class="section level1">
<h1>2015</h1>
<div id="workshop-on-metagenomics-harvard" class="section level2">
<h2>Workshop on Metagenomics (Harvard)</h2>
<ul>
<li><a href="http://evomics.org/workshops/harvard-university-workshops/2015-harvard-university-workshop-on-metagenomics/">Harvard University Workshop on Metagenomics (2015 Sept 14 - 18)</a></li>
</ul>
<p>A week-long workshop focused exclusively on the analysis of metagenomic data. This intensive, hands-on course will feature morning and afternoon sessions taught by expert faculty and technical assistants from the Broad, Ragon Institute, Harvard School of Public Health, Stanford University and Washington University, St. Louis. All participants will acquire fundamental computational skills, appropriate study design, data interpretation, and overall literacy in bioinformatics.</p>
<p>When: 2015 14-18, September : Morning Sessions 9:00 – 12:00 | Afternoon Sessions 2:00 – 5:00</p>
<p>Where: Joseph B. Martin Conference Center Harvard Medical School 77 Avenue Louis Pasteur Boston, MA 02115.</p>
<p> </p>
</div>
<div id="sismid-workshop-2015" class="section level2">
<h2>SISMID Workshop, 2015</h2>
<ul>
<li><a href="http://www.biostat.washington.edu/suminst/sismid/modules">Summer Institute in Statistics, Modeling in Infectious Diseases (SISMID): Metagenomics Module</a></li>
</ul>
<p> </p>
</div>
<div id="stamps-workshop-2015" class="section level2">
<h2>STAMPS Workshop, 2015</h2>
<ul>
<li><a href="https://stamps.mbl.edu/index.php/Main_Page">Strategies and Techniques for Analyzing Microbial Population Structure (STAMPS)</a></li>
</ul>
<p>Course Date: August 5 - August 15, 2015</p>
<p>Twitter hash: #STAMPS15</p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
