<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title></title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<link rel="stylesheet"
      href="site_libs/highlight/default.css"
      type="text/css" />
<script src="site_libs/highlight/highlight.js"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>

<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script src="site_libs/navigation-1.0/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">phyloseq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    phyloseq
  </a>
</li>
<li>
  <a href="install.html">
    <span class="fa fa-gear"></span>
     
    Installation
  </a>
</li>
<li>
  <a href="http://joey711.github.io/phyloseq-extensions">
    <span class="fa fa-gear"></span>
     
    Extensions
  </a>
</li>
<li>
  <a href="https://github.com/joey711/phyloseq/issues">
    <span class="fa fa-medkit"></span>
     
    Issues
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-newspaper-o"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0061217">phyloseq in Bioconductor</a>
    </li>
    <li>
      <a href="http://bioinformatics.oxfordjournals.org/content/early/2014/09/26/bioinformatics.btu616">Shiny-phyloseq</a>
    </li>
    <li>
      <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003531">Rarefying Inadmissible</a>
    </li>
    <li>
      <a href="http://f1000research.com/articles/5-1492/v1">Full Stack Workflow</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-university"></span>
     
    Courses
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
        Upcoming Courses
         
        <span class="caret"></span>
      </a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="http://www.biostat.washington.edu/suminst/sismid2016/modules/MD1614">SISMID 2016 Module 14 Intro. to Metagenomic Analysis</a>
        </li>
        <li>
          <a href="http://evomics.org/workshops/workshop-on-genomics/">Workshop on Genomics, Český Krumlov, Czech Republic (2017 January)</a>
        </li>
      </ul>
    </li>
    <li class="dropdown">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
        Previous Courses
         
        <span class="caret"></span>
      </a>
      <ul class="dropdown-menu" role="menu">
        <li class="dropdown-header">(2014 - present) Summer Institute in Statistics, Modeling in Infectious Diseases (SISMID): Metagenomics Module</li>
        <li>
          <a href="http://evomics.org/workshops/workshop-on-genomics/2016-workshop-on-genomics-cesky-krumlov/">Workshop on Genomics, Cesky Krumlov, Czech Republic (2016 Jan 10 - 23)</a>
        </li>
        <li>
          <a href="https://stamps.mbl.edu/index.php/Main_Page">Strategies and Techniques for Analyzing Microbial Population Structure (STAMPS)</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-university"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="import-data.html">Data Import</a>
    </li>
    <li>
      <a href="Example-Data.html">Example Data</a>
    </li>
    <li>
      <a href="merge.html">Merge</a>
    </li>
    <li>
      <a href="preprocess.html">Preprocess</a>
    </li>
    <li>
      <a href="distance.html">Distances</a>
    </li>
    <li>
      <a href="gap-statistic.html">Gap Statistic</a>
    </li>
    <li>
      <a href="plot_ordination-examples.html">Plot Ordination</a>
    </li>
    <li>
      <a href="plot_richness-examples.html">Plot Alpha Diversity</a>
    </li>
    <li>
      <a href="plot_tree-examples.html">Plot Tree</a>
    </li>
    <li>
      <a href="plot_bar-examples.html">Plot Bars</a>
    </li>
    <li>
      <a href="plot_heatmap-examples.html">Plot Heatmap</a>
    </li>
    <li>
      <a href="plot_network-examples.html">Plot Network</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/phyloseq/inst/doc/phyloseq-FAQ.html">
    <span class="fa fa-question fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/joey711">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">




</div>


<p><link href="http://joey711.github.com/phyloseq/markdown.css" rel="stylesheet"></link></p>
<div id="download-microbio.meqiime-datasets" class="section level1">
<h1>Download microbio.me/qiime datasets</h1>
<p><a href="http://joey711.github.io/phyloseq/download-microbio.me.html">Originally hosted here</a></p>
<p>Here is an example accessing microbiome datasets from a public repository, using entirely R code. I haven’t yet figured out how to list the available studies at The <a href="http://www.microbio.me/qiime/index.psp">microbio.me/qiime</a> from within R, but I have provided illustrate instructions for finding details about studies you might want to download and explore, and some example FTP addresses that I used after doing just that. Note that you likely need to create an account at The <a href="http://www.microbio.me/qiime/index.psp">microbio.me/qiime</a> in order to explore the studies they have available for download.</p>
<div id="microbio.meqiime-microbio_me_qiime-in-phyloseq" class="section level2">
<h2>microbio.me/qiime –&gt; <em>microbio_me_qiime</em> in phyloseq</h2>
<p>The <code>microbio_me_qiime</code> interface is a means to import data directly from the <a href="http://www.microbio.me/qiime/index.psp">microbio.me/qiime</a> data repository with only a small quantity of effort.</p>
<p>There are datasets posted on <a href="http://www.microbio.me/qiime/index.psp">microbio.me/qiime</a>, and these usually have the “raw” OTU clustered data hosted at an FTP address. I’m not sure of a “nice” way to explor the details of the different studies using the FTP address directly, but I can post a few hand-picked datasets and import them all.</p>
<p>First, <a href="http://www.microbio.me/qiime/index.psp">create an account and login</a>. Here is what the frontpage should look like:</p>
<p><img src="http://joey711.github.io/phyloseq/microbio-me-login.png" width="750px" /></p>
<p>Once you have logged-in, you will see a different screen. Click on “Get Study Summary and Raw Data”, which will be a link near the top of the page (highlighted at the bottom of the following clipped page):</p>
<p><img src="http://joey711.github.io/phyloseq/microbio-me-download-data.png" width="400px" /></p>
<p>And once you’re at this “raw data” page, you will see a box below “Available Studies”. The box is a large scrollable list, and the page below it will start off blank:</p>
<p><img src="http://joey711.github.io/phyloseq/microbio-me-data-databox.png" width="250px" /></p>
<p>But once you click on something, details and additional links for that particular study will pop up. For example, I clicked on a version of the “Global Patterns” dataset in the box, which exposed a page below with further details about the study and links for data. If available, there is an FTP link called “Sequences, Mapping and OTU Table”, which can be accessed directly from within R using the code provided farther down this page. The link itself is highlighted on this screenshot:</p>
<p><img src="http://joey711.github.io/phyloseq/microbio-me-gpdata.png" width="250px" /></p>
<hr />
</div>
</div>
<div id="download-parse-datasets" class="section level1">
<h1>Download, parse datasets</h1>
<p>Load phyloseq package… and ggplot2 for plotting at the end…</p>
<pre class="r"><code>library(&quot;phyloseq&quot;)
packageVersion(&quot;phyloseq&quot;)</code></pre>
<pre><code>## [1] &#39;1.5.21&#39;</code></pre>
<pre class="r"><code>library(&quot;ggplot2&quot;)
packageVersion(&quot;ggplot2&quot;)</code></pre>
<pre><code>## [1] &#39;0.9.3.1&#39;</code></pre>
<pre class="r"><code>theme_set(theme_bw())</code></pre>
<p>I previously defined a function here “on the fly” to download and parse data from the FTP server for <code>microbio.me</code>, and also download, unpack and import into R/<a href="http://joey711.github.io/phyloseq/">phyloseq</a>. A more complete version of this functionality is now included in phyloseq, and this page now demonstrates how that can work.</p>
<p>The main argument to this function identifies which study in the microbio.me/qiime database that you want to download and import. Here are several hand-picked examples using either just the study ID number.</p>
<pre class="r"><code># Restroom Surfaces - Flores_restroom_surface_biogeography
restroom = microbio_me_qiime(1335)</code></pre>
<pre><code>## Found biom-format file, now parsing it... 
## Done parsing biom... 
## Importing Sample Metdadata from mapping file...
## Merging the imported objects... 
## Successfully merged, phyloseq-class created. 
##  Returning...</code></pre>
<pre class="r"><code># smokers - Charlson_cigarette_smokers
smokers = microbio_me_qiime(524)</code></pre>
<pre><code>## Found biom-format file, now parsing it... 
## Done parsing biom... 
## Importing Sample Metdadata from mapping file...
## Merging the imported objects... 
## Successfully merged, phyloseq-class created. 
##  Returning...</code></pre>
<pre class="r"><code># rs - resistant starches - Martinez_Resistant_starches
rs = microbio_me_qiime(495)</code></pre>
<pre><code>## Found biom-format file, now parsing it... 
## Done parsing biom... 
## Importing Sample Metdadata from mapping file...
## Merging the imported objects... 
## Successfully merged, phyloseq-class created. 
##  Returning...</code></pre>
<pre class="r"><code># abt - antibiotic timecourse - Relman_antibiotic_timeseries
abt = microbio_me_qiime(494)</code></pre>
<pre><code>## Found biom-format file, now parsing it... 
## Done parsing biom... 
## Importing Sample Metdadata from mapping file...
## Merging the imported objects... 
## Successfully merged, phyloseq-class created. 
##  Returning...</code></pre>
<pre class="r"><code># vagina - vaginal microbiome - Ravel_reproductive_women_vagina
vagina = microbio_me_qiime(509)</code></pre>
<pre><code>## Found biom-format file, now parsing it... 
## Done parsing biom... 
## Importing Sample Metdadata from mapping file...
## Merging the imported objects... 
## Successfully merged, phyloseq-class created. 
##  Returning...</code></pre>
<pre class="r"><code># palms - Fierer_undergraduate_palms
palms = microbio_me_qiime(317)</code></pre>
<pre><code>## Found biom-format file, now parsing it... 
## Done parsing biom... 
## Importing Sample Metdadata from mapping file...
## Merging the imported objects... 
## Successfully merged, phyloseq-class created. 
##  Returning...</code></pre>
<p>Here is an example downloading a version of the Global Patterns dataset using the full URL (ftp) address:</p>
<pre class="r"><code># # Global Patterns - CaporasoIlluminaPNAS2011_3prime
gpftp = &quot;ftp://thebeast.colorado.edu/pub/QIIME_DB_Public_Studies/study_721_split_library_seqs_and_mapping.tgz&quot;
gp = microbio_me_qiime(gpftp)</code></pre>
<pre><code>## Found biom-format file, now parsing it... 
## Done parsing biom... 
## Importing Sample Metdadata from mapping file...
## Merging the imported objects... 
## Successfully merged, phyloseq-class created. 
##  Returning...</code></pre>
<div id="some-minimal-preprocessing" class="section level2">
<h2>Some minimal <a href="http://joey711.github.io/phyloseq/preprocessing.html">preprocessing</a></h2>
<p>Remove “empty” OTUs and samples from each dataset. Note that I’ve chosen to keep only samples that have more than 100 total reads</p>
<pre class="r"><code>dlist = list(restroom, smokers, rs, abt, vagina, palms)
dlist = lapply(dlist, function(physeq) {
    physeq = prune_taxa(taxa_sums(physeq) &gt; 0, physeq)
    physeq = prune_samples(sample_sums(physeq) &gt; 100, physeq)
})</code></pre>
</div>
<div id="plots" class="section level2">
<h2>Plots</h2>
<p>And some plots, because humans like images and it also helps demonstrate that the data is now directly available for analysis.</p>
<div id="richness-plots" class="section level3">
<h3><a href="http://joey711.github.io/phyloseq/plot_richness-examples.html">Richness plots</a></h3>
<pre class="r"><code>plot_richness(smokers, x = &quot;AGE&quot;, color = &quot;SEX&quot;, shape = &quot;BODY_PRODUCT&quot;)</code></pre>
<div class="figure">
<img src="figure/richness-plots1.png" alt="plot of chunk richness-plots" />
<p class="caption">plot of chunk richness-plots</p>
</div>
<pre class="r"><code>plot_richness(vagina, x = &quot;NUGENT_SCORE&quot;, color = &quot;ETHNICITY&quot;) + stat_smooth(method = lm)</code></pre>
<div class="figure">
<img src="figure/richness-plots2.png" alt="plot of chunk richness-plots" />
<p class="caption">plot of chunk richness-plots</p>
</div>
<pre class="r"><code>plot_richness(vagina, x = &quot;PH&quot;, color = &quot;ETHNICITY&quot;) + stat_smooth(method = lm)</code></pre>
<div class="figure">
<img src="figure/richness-plots3.png" alt="plot of chunk richness-plots" />
<p class="caption">plot of chunk richness-plots</p>
</div>
</div>
<div id="heatmap" class="section level3">
<h3><a href="http://joey711.github.io/phyloseq/plot_heatmap-examples.html">Heatmap</a></h3>
<pre class="r"><code>plot_heatmap(smokers)</code></pre>
<div class="figure">
<img src="figure/heatmap.png" alt="plot of chunk heatmap" />
<p class="caption">plot of chunk heatmap</p>
</div>
</div>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
