<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Future Development of phyloseq</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}
</style>





</head>

<body>
<p><link href="markdown.css" rel="stylesheet"></link></p>

<h1>Future Development of phyloseq</h1>

<p>Additional development of phyloseq is ongoing, the details of which are documented heavily at <a href="https://github.com/joey711/phyloseq/issues">the phyloseq issues tracker</a>.</p>

<p>In broader strokes, the near-term plans include:</p>

<ol>
<li><p>The &ldquo;compartmentalization&rdquo; of the data infrastructure portion of the phyloseq package into a separate Bioconductor package called <a href="https://github.com/joey711/phyloseq/issues/102">phyloseqBase</a>. This may reduce the number of dependencies implied when reusing the data tools of phyloseq for other Bioconductor/R packages.</p></li>
<li><p>The addition to phyloseqBase of interfaces to public repositories for integrated microbiome census data; for instance, <a href="http://metagenomics.anl.gov/">MG-RAST</a> and <a href="http://microbio.me/qiime">microbe.me</a>.</p></li>
<li><p>More advanced tree-based OTU clustering, leveraging tools already available in R.</p></li>
<li><p>The development and calibration of additional tools in phyloseqBase for formal preprocessing, filtering, normalization, shrinkage and variance stabilization (<a href="http://www.nature.com/nrg/journal/v7/n1/full/nrg1749.html">e.g. Allison 2006</a>) </p></li>
<li><p>Additional structured wrapping tools to the <a href="http://cran.r-project.org/web/packages/ade4/index.html">ade4</a> and <a href="http://cran.r-project.org/web/packages/vegan/index.html">vegan</a> packages, with the most commonly-used tools given specific wrapping functions in phyloseq or phyloseqBase.</p></li>
<li><p>Structures for additional data components, e.g. mass spec and expression data.</p></li>
<li><p>Animated ordinations for time-series (and analogous) data. See beta-version support package: <a href="https://github.com/joey711/animate.phyloseq">animate.phyloseq</a></p></li>
<li><p>The compilation of a <a href="http://www.bioconductor.org/packages/release/data/experiment/">Bioconductor data package</a> (&ldquo;phyloseqData&rdquo;) that includes many key published datasets already imported as separate &ldquo;phyloseq&rdquo; instances and available through R&#39;s <code>data</code> interface.</p></li>
<li><p><strong>Big(ger)</strong> Data
The firehose of new-gen sequencing data is making possible &ldquo;big&rdquo; datasets in this realm. For example, <a href="http://joey711.github.com/phyloseq-demo/HMP_import_example.html">the demo on importing the Human Microbiome Project data into R</a> takes a considerable amount of time to run on a typical desktop/laptop and may push some less powerful machines to their limit. And that&#39;s just data import. We are considering some of the best approaches to help a tool like phyloseq address computational issues that are arising from dealing with this data of this size, without compromising some of the other features (interactivity, reproducibility, connection with existing R tools). Some promising tools already available for R that might help include:</p></li>
</ol>

<h5>Sparse matrix classes</h5>

<p>Like in <a href="http://cran.r-project.org/web/packages/Matrix/index.html">the Matrix package</a>. This mainly applies to in-RAM computations, especially when the full matrix is actually needed. In principle, this might only apply be required to represent the preprocessed data, which could still be sparse.</p>

<h5>Store full dataset in a database</h5>

<p>Some candidate packages are:</p>

<ul>
<li><p><a href="http://cran.r-project.org/web/packages/hdf5/index.html">the hdf5 package</a></p></li>
<li><p><a href="http://cran.r-project.org/web/packages/ff/index.html">the ff package</a></p></li>
</ul>

<hr/>

<h3>Other tutorial pages for the phyloseq package:</h3>

<h4><a href="distance.html">distance</a></h4>

<h4><a href="download-microbio.me-frag.html">download-microbio.me-frag</a></h4>

<h4><a href="download-microbio.me.html">download-microbio.me</a></h4>

<h4><a href="Example-Data.html">Example-Data</a></h4>

<h4><a href="future-devel.html">future-devel</a></h4>

<h4><a href="import-data.html">import-data</a></h4>

<h4><a href="index.html">index</a></h4>

<h4><a href="install.html">install</a></h4>

<h4><a href="merge.html">merge</a></h4>

<h4><a href="plot_bar-examples.html">plot_bar-examples</a></h4>

<h4><a href="plot_heatmap-examples.html">plot_heatmap-examples</a></h4>

<h4><a href="plot_network-examples.html">plot_network-examples</a></h4>

<h4><a href="plot_ordination-examples.html">plot_ordination-examples</a></h4>

<h4><a href="plot_richness-examples.html">plot_richness-examples</a></h4>

<h4><a href="plot_tree-examples.html">plot_tree-examples</a></h4>

<h4><a href="preprocess.html">preprocess</a></h4>

<h4><a href="subset_ord_plot-examples.html">subset_ord_plot-examples</a></h4>

<h4><a href="tutorials-index.html">tutorials-index</a></h4>

</body>

</html>
