<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="phyloseq : import, share, and analyze phylogenetic sequencing data using R" />

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>phyloseq</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/joey711/phyloseq">View on GitHub</a>

          <h1 id="project_title">phyloseq</h1>
          <h2 id="project_tagline">import, share, and analyze microbiome census data using R</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/joey711/phyloseq/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/joey711/phyloseq/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h1>phyloseq</h1>


<h3>Installation</h3>

<p>Please see the <a href="http://joey711.github.com/phyloseq/install">phyloseq installation instructions</a>.</p>
      
<h3>Tutorials</h3>

<p>The complete list of <a
      href="http://joey711.github.com/phyloseq/tutorials-index">phyloseq
      frontpage tutorials</a>.</p>

<h3>phyloseq-demo Repository</h3>
<p><a href="https://github.com/joey711/phyloseq-demo">The phyloseq-demo repository</a> contains many additional documentation materials created to support workshops on the use of phyloseq in microbiome research. These materials are available under an open-access copyright license.</p>

<h3>Article on Microbiome Normalization and Differential Abundance Testing</h3>

<p>McMurdie and Holmes <a href="http://arxiv.org/abs/1310.0424">Waste Not, Want Not: Why Rarefying Microbiome Data is Inadmissible</a></p>

<p> (2014) <a href="http://www.ploscompbiol.org/">PLoS Computational
      Biology</a> <b>in press</b> </p>
      
<p> (2013) Available ahead of acceptance.
<a href="http://arxiv.org/pdf/1310.0424v2.pdf">PDF version 2 (12 Dec 2013)</a>,
<a href="http://arxiv.org/pdf/1310.0424v1.pdf">PDF version 1 (1 Oct 2013)</a></p>

<h3>Latest Peer-Reviewed Article about phyloseq</h3>
<p>McMurdie and Holmes (2013) phyloseq: An R Package for Reproducible Interactive Analysis and
      Graphics of Microbiome Census Data. PLoS ONE. 8(4):e61217</p>
<a href="http://dx.plos.org/10.1371/journal.pone.0061217">Free, open-access online at PLoS ONE</a>
      
<h2>Description</h2>

<p>The analysis of microbiological communities brings many challenges: the integration of many different types of data with methods from ecology, genetics, phylogenetics, network analysis, visualization and testing. The data itself may originate from widely different sources, such as the microbiomes of humans, soils, surface and ocean waters, wastewater treatment plants, industrial facilities, and so on; and as a result, these varied sample types may have very different forms and scales of related data that is extremely dependent upon the experiment and its question(s). The phyloseq package is a tool to import, store, analyze, and graphically display complex phylogenetic sequencing data that has already been clustered into Operational Taxonomic Units (OTUs), especially when there is associated sample data, phylogenetic tree, and/or taxonomic assignment of the OTUs. This package leverages many of the tools available in R for ecology and phylogenetic analysis (vegan, ade4, ape, picante), while also using advanced/flexible graphic systems (ggplot2) to easily produce publication-quality graphics of complex phylogenetic data. phyloseq uses a specialized system of S4 classes to store all related phylogenetic sequencing data as single experiment-level object, making it easier to share data and reproduce analyses. In general, phyloseq seeks to facilitate the use of R for efficient interactive and reproducible analysis of OTU-clustered high-throughput phylogenetic sequencing data.</p>

<p>More concretely, <a href="http://joey711.github.com/phyloseq/">phyloseq</a> provides:</p>

<ul>
<li>
<p><a href="http://joey711.github.com/phyloseq/import-data">Import</a> abundance and related data from popular OTU clustering pipelines:</p>

<ul>
<li> See the tutorial page on <a
  href="http://joey711.github.com/phyloseq/import-data">importing data
  with phyloseq</a> for more
  details</li>
<li>See the <a
  href="http://joey711.github.io/phyloseq/download-microbio.me.html">microbio_me_qiime
  tutorial page</a> on importing directly from the <a href="http://www.microbio.me/qiime/">microbio.me/qiime</a> data server </li>
<li>
Works with <a href="http://qiime.org/">QIIME</a>, <a href="http://www.mothur.org/">mothur</a>, <a href="http://www.qiime.org/svn_documentation/documentation/biom_format.html">BIOM</a>, <a href="http://pyrotagger.jgi-psf.org/cgi-bin/index.pl">PyroTagger</a>, <a href="http://pyro.cme.msu.edu/">RDP</a>. </li>
<li>Additional importers planned for <a href="http://metagenomics.anl.gov/">MG-RAST</a>, <a href="http://clovr.org/methods/clovr-16s/">CLoVR-16S</a>, <a href="http://genboree.org/theCommons/projects/pub-gen-microbiome">Genboree Microbiome</a>, <a href="http://www.microgator.org/pangea/">PANGEA</a>, and others.</li>
</ul>
</li>
<li>
<p>Convenience analysis wrappers for common analysis tasks, e.g.</p>

<ul>
<li>
<a href="http://joey711.github.com/phyloseq/distance">distance</a>  --&gt; 44 distance methods supported (UniFrac, Jensen-Shannon, etc)</li>
<li>
<a href="https://github.com/joey711/phyloseq/wiki/ordinate">ordinate</a>  --&gt; many supported methods, including constrained methods</li>
</ul>
</li>
<li>
<p>Powerful, flexible custom plot methods using ggplot2 for rapid, convenient exploratory analysis.</p>

<ul>
<li><a href="http://joey711.github.com/phyloseq/plot_heatmap-examples">plot_heatmap</a></li>
<li><a href="http://joey711.github.com/phyloseq/plot_tree-examples">plot_tree</a></li>
<li><a href="http://joey711.github.com/phyloseq/plot_ordination-examples">plot_ordination</a></li>
<li><a href="http://joey711.github.com/phyloseq/plot_network-examples">plot_network</a></li>
<li><a href="http://joey711.github.com/phyloseq/plot_richness-examples">plot_richness</a></li>
<li><a href="http://joey711.github.com/phyloseq/plot_bar-examples">plot_bar</a></li>
</ul>
</li>
<li><p>Modular, customizable <a
href="http://joey711.github.com/phyloseq/preprocess">preprocessing
functions</a> supporting fully reproducible work.</p></li>
<li><p>Functions for <a
href="http://joey711.github.com/phyloseq/merge">merging data</a> based
on OTU/sample variables, and for supporting manually-imported data.</p></li>
<li><p>Native, parallelized implementation of <a href="https://github.com/joey711/phyloseq/wiki/Fast-Parallel-UniFrac">UniFrac</a> distance calculations.</p></li>
<li><p>Multiple testing methods specific to high-throughput phylogenetic sequencing data.</p></li>
<li><p>Examples for analysis and graphics using real published data.</p></li>
<li><p>Some simple tree-based OTU clustering for building a combined OTU/tree data object "from scratch".</p></li>
</ul><p>Additional features planned for the near-term are outlined at
<a
href="http://joey711.github.com/phyloseq/future-devel">the
future development page</a>.</p>

<p>For news related to the latest development version, see:</p>

<p><a href="https://github.com/joey711/phyloseq/blob/master/inst/NEWS">phyloseq NEWS</a></p>

<p>The phyloseq package is actively and openly developed on GitHub:</p>
<a href="https://github.com/joey711/phyloseq">the phyloseq development
page</a></p>

<p>I make lots of effort to cite/attribute author contributions within official package documentation, citations, and anywhere else it is appropriate. Please feel free to fork and contribute!</p>
      
<h3>Wiki</h3>

<p>Post your own documentation at the <a href="https://github.com/joey711/phyloseq/wiki">phyloseq wiki</a>.</p>

<h2>Workshop</h2>

<p><a href="https://secure.bioconductor.org/BioC2012/labs.php">A workshop session</a> regarding the use of <a href="http://joey711.github.com/phyloseq/">the phyloseq package</a> in amplicon sequencing analysis was provided at the</p>

<p><a href="https://secure.bioconductor.org/BioC2012/">Bioconductor Workshop (BioC 2012)</a></p>

<p>on July 24-25, 2012 </p>

<p>At the Fred Hutchinson Cancer Research Center - Seattle, WA. </p>

<p>The materials for the workshop are provided under copyright license through <a href="https://github.com/joey711/phyloseq-demo">the phyloseq-demo repository</a>, which we will continue to revise and update as we continue to improve the documentation for the phyloseq package. </p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">phyloseq maintained by <a href="https://github.com/joey711">joey711</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

              <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-32799541-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>


  </body>
</html>
