<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Merge Data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<link href="site_libs/ionicons-2.0.1/css/ionicons.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}

.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">phyloseq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="install.html">
    <span class="ion ion-android-archive"></span>
     
    Installation
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-newspaper-o"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0061217">phyloseq in Bioconductor</a>
    </li>
    <li>
      <a href="http://bioinformatics.oxfordjournals.org/content/early/2014/09/26/bioinformatics.btu616">Shiny-phyloseq</a>
    </li>
    <li>
      <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003531">Rarefying Inadmissible</a>
    </li>
    <li>
      <a href="http://f1000research.com/articles/5-1492/v1">Full Stack Workflow</a>
    </li>
    <li>
      <a href="https://www.nature.com/articles/nmeth.3869">DADA2</a>
    </li>
    <li>
      <a href="https://www.nature.com/articles/ismej2017119">Amplicon Sequence Variants &gt; OTUs</a>
    </li>
  </ul>
</li>
<li>
  <a href="courses.html">
    <span class="fa fa-graduation-cap"></span>
     
    Courses
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-university"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Basics</li>
    <li>
      <a href="import-data.html">Data Import</a>
    </li>
    <li>
      <a href="Example-Data.html">Example Data</a>
    </li>
    <li>
      <a href="merge.html">Merge</a>
    </li>
    <li>
      <a href="preprocess.html">Preprocess</a>
    </li>
    <li>
      <a href="distance.html">Distances</a>
    </li>
    <li>
      <a href="gap-statistic.html">Gap Statistic</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Workflows</li>
    <li>
      <a href="http://benjjneb.github.io/dada2/tutorial.html">DADA2</a>
    </li>
    <li>
      <a href="http://f1000research.com/articles/5-1492/v1">Full Stack</a>
    </li>
    <li>
      <a href="http://joey711.github.io/phyloseq-demo/HMP_import_example.html">HMP-1</a>
    </li>
    <li>
      <a href="http://joey711.github.io/phyloseq-demo/Restroom-Biogeography.html">Restroom Biogeography</a>
    </li>
    <li>
      <a href="http://joey711.github.io/phyloseq-demo/unifrac.html">UniFrac</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Graphics</li>
    <li>
      <a href="plot_ordination-examples.html">Plot Ordination</a>
    </li>
    <li>
      <a href="plot_richness-examples.html">Plot Alpha Diversity</a>
    </li>
    <li>
      <a href="plot_tree-examples.html">Plot Tree</a>
    </li>
    <li>
      <a href="plot_bar-examples.html">
        <span class="fa fa-bar-chart"></span>
         
        Plot Bars
      </a>
    </li>
    <li>
      <a href="plot_heatmap-examples.html">
        <span class="fa fa-qrcode"></span>
         
        Plot Heatmap
      </a>
    </li>
    <li>
      <a href="plot_network-examples.html">
        <span class="ion ion-network"></span>
         
        Plot Network
      </a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Differential Abundance</li>
    <li>
      <a href="http://joey711.github.io/phyloseq-extensions/DESeq2.html">DESeq2 (Recommended)</a>
    </li>
    <li>
      <a href="http://joey711.github.io/phyloseq-extensions/edgeR.html">edgeR</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/joey711/phyloseq/issues">
    <span class="fa fa-medkit"></span>
     
    Issues
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/phyloseq/inst/doc/phyloseq-FAQ.html">
    <span class="fa fa-question"></span>
     
    FAQ
  </a>
</li>
<li>
  <a href="https://github.com/joey711/phyloseq">
    <span class="fa fa-github-alt"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-twitter"></span>
     
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://twitter.com/joey711">
        <span class="fa fa-twitter"></span>
         
      </a>
    </li>
    <li>
      <a href="https://www.linkedin.com/in/joey711">
        <span class="fa fa-linkedin-square"></span>
         
      </a>
    </li>
    <li>
      <a href="http://stackoverflow.com/users/935950/paul-mcmurdie">
        <span class="fa fa-stack-overflow"></span>
         
      </a>
    </li>
    <li>
      <a href="https://scholar.google.com/citations?user=nNQ1LtMAAAAJ">
        <span class="ai ai-google-scholar-square"></span>
         
      </a>
    </li>
    <li>
      <a href="https://www.researchgate.net/profile/Paul_Mcmurdie">
        <span class="ai ai-researchgate-square"></span>
         
      </a>
    </li>
    <li>
      <a href="https://github.com/joey711/phyloseq">
        <span class="fa fa-github-alt"></span>
         
      </a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Merge Data</h1>

</div>


<p><a href="http://joey711.github.com/phyloseq/">The phyloseq project</a> includes support for two completely different categories of <strong>merging</strong> data objects.</p>
<ul>
<li><strong>Merging the OTUs or samples in a phyloseq object</strong>, based upon a taxonomic or sample variable: <code>merge_samples()</code> <code>merge_taxa()</code></li>
</ul>
<p>Merging OTU or sample indices based on variables in the data can be a useful means of reducing noise or excess features in an analysis or graphic. Some examples might be to merge the samples in a dataset that are from the same environment, or orthogonally, to merge OTUs that are from the same taxonomic genera. Often this takes the form of a table-join, where non-matching keys are omitted in the result. In this case, keys are Sample IDs or Taxa IDs, respectively.</p>
<ul>
<li><strong>Merging two or more data objects</strong> that come from the same experiment, so that their data becomes part of the same phyloseq object: <code>merge_phyloseq()</code></li>
</ul>
<p>Merging separate data objects is especially useful for manually-imported data objects, especially when one of the data objects already has more than one component and so is a phyloseq-class. While the first category of merging functions is useful for direct manipulations of the data for analytical purposes, <code>merge_phyloseq</code> is a convenience/support tool to help get your data into the right format.</p>
<div id="examples" class="section level1">
<h1>Examples</h1>
<div id="merge_samples" class="section level2">
<h2>merge_samples</h2>
<p><code>merge_samples</code> can be very useful if you’d like to see what happens to an analysis if you remove the indivual effects between replicates or between samples from a particular explanatory variable. With the <code>merge_samples</code> function, the abundance values of merged samples are summed, so make sure to do any <a href="http://joey711.github.com/phyloseq/preprocess">preprocessing</a> to account for differences in sequencing effort before merging or you will achieve a sequencing-effort-weighted average (which you may want, but keep in mind).</p>
<p>Let’s first remove unobserved OTUs (sum 0 across all samples), and add a human-associated variable with which to organize data later in the plots.</p>
<pre class="r"><code>library(&quot;phyloseq&quot;); packageVersion(&quot;phyloseq&quot;)</code></pre>
<pre><code>## [1] &#39;1.22.3&#39;</code></pre>
<p>Load data, remove empty samples, add a new <code>sample_data</code> variable to the dataset.</p>
<pre class="r"><code>data(GlobalPatterns)
GP = GlobalPatterns
GP = prune_taxa(taxa_sums(GlobalPatterns) &gt; 0, GlobalPatterns)
humantypes = c(&quot;Feces&quot;, &quot;Mock&quot;, &quot;Skin&quot;, &quot;Tongue&quot;)
sample_data(GP)$human &lt;- get_variable(GP, &quot;SampleType&quot;) %in% humantypes</code></pre>
<p>Now on to the merging examples</p>
<pre class="r"><code>mergedGP = merge_samples(GP, &quot;SampleType&quot;)
SD = merge_samples(sample_data(GP), &quot;SampleType&quot;)
print(SD[, &quot;SampleType&quot;])</code></pre>
<pre><code>##                    SampleType
## Feces                       1
## Freshwater                  2
## Freshwater (creek)          3
## Mock                        4
## Ocean                       5
## Sediment (estuary)          6
## Skin                        7
## Soil                        8
## Tongue                      9</code></pre>
<pre class="r"><code>print(mergedGP)</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 18988 taxa and 9 samples ]
## sample_data() Sample Data:       [ 9 samples by 8 sample variables ]
## tax_table()   Taxonomy Table:    [ 18988 taxa by 7 taxonomic ranks ]
## phy_tree()    Phylogenetic Tree: [ 18988 tips and 18987 internal nodes ]</code></pre>
<pre class="r"><code>sample_names(GP)</code></pre>
<pre><code>##  [1] &quot;CL3&quot;      &quot;CC1&quot;      &quot;SV1&quot;      &quot;M31Fcsw&quot;  &quot;M11Fcsw&quot;  &quot;M31Plmr&quot; 
##  [7] &quot;M11Plmr&quot;  &quot;F21Plmr&quot;  &quot;M31Tong&quot;  &quot;M11Tong&quot;  &quot;LMEpi24M&quot; &quot;SLEpi20M&quot;
## [13] &quot;AQC1cm&quot;   &quot;AQC4cm&quot;   &quot;AQC7cm&quot;   &quot;NP2&quot;      &quot;NP3&quot;      &quot;NP5&quot;     
## [19] &quot;TRRsed1&quot;  &quot;TRRsed2&quot;  &quot;TRRsed3&quot;  &quot;TS28&quot;     &quot;TS29&quot;     &quot;Even1&quot;   
## [25] &quot;Even2&quot;    &quot;Even3&quot;</code></pre>
<pre class="r"><code>sample_names(mergedGP)</code></pre>
<pre><code>## [1] &quot;Feces&quot;              &quot;Freshwater&quot;         &quot;Freshwater (creek)&quot;
## [4] &quot;Mock&quot;               &quot;Ocean&quot;              &quot;Sediment (estuary)&quot;
## [7] &quot;Skin&quot;               &quot;Soil&quot;               &quot;Tongue&quot;</code></pre>
<pre class="r"><code>identical(SD, sample_data(mergedGP))</code></pre>
<pre><code>## [1] TRUE</code></pre>
<p>As emphasized earlier, the OTU abundances of merged samples are summed. Let’s investigate this ourselves looking at just the top10 most abundance OTUs.</p>
<pre class="r"><code>OTUnames10 = names(sort(taxa_sums(GP), TRUE)[1:10])
GP10  = prune_taxa(OTUnames10,  GP)
mGP10 = prune_taxa(OTUnames10, mergedGP)
ocean_samples = sample_names(subset(sample_data(GP), SampleType==&quot;Ocean&quot;))
print(ocean_samples)</code></pre>
<pre><code>## [1] &quot;NP2&quot; &quot;NP3&quot; &quot;NP5&quot;</code></pre>
<pre class="r"><code>otu_table(GP10)[, ocean_samples]</code></pre>
<pre><code>## OTU Table:          [10 taxa and 3 samples]
##                      taxa are rows
##         NP2   NP3   NP5
## 329744   91   126   120
## 317182 3148 12370 63084
## 549656 5045 10713  1784
## 279599  113   114   126
## 360229   16    83   786
## 94166    49   128   709
## 550960   11    86    65
## 158660   13    39    28
## 331820   24   101   105
## 189047    4    33    29</code></pre>
<pre class="r"><code>rowSums(otu_table(GP10)[, ocean_samples])</code></pre>
<pre><code>## 329744 317182 549656 279599 360229  94166 550960 158660 331820 189047 
##    337  78602  17542    353    885    886    162     80    230     66</code></pre>
<pre class="r"><code>otu_table(mGP10)[&quot;Ocean&quot;, ]</code></pre>
<pre><code>## OTU Table:          [10 taxa and 1 samples]
##                      taxa are columns
##       329744 317182 549656 279599 360229 94166 550960 158660 331820 189047
## Ocean    337  78602  17542    353    885   886    162     80    230     66</code></pre>
<p>Let’s look at the merge graphically between two <a href="http://joey711.github.com/phyloseq/plot_richness-examples">richness estimate summary plots</a>. Note that we are ignoring our own advice and not-performing any preprocessing before this merge. Since these are richness estimates we are probably safe.</p>
<pre class="r"><code>plot_richness(GP, &quot;human&quot;, &quot;SampleType&quot;, title=&quot;unmerged&quot;)</code></pre>
<pre><code>## Warning: Removed 130 rows containing missing values (geom_errorbar).</code></pre>
<p><img src="merge_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>The merge can do some weird things to sample variables. Let’s re-add these variables to the <code>sample_data</code> before we plot.</p>
<pre class="r"><code>sample_data(mergedGP)$SampleType = sample_names(mergedGP)
sample_data(mergedGP)$human = sample_names(mergedGP) %in% humantypes
plot_richness(mergedGP, &quot;human&quot;, &quot;SampleType&quot;, title=&quot;merged&quot;)</code></pre>
<pre><code>## Warning: Removed 45 rows containing missing values (geom_errorbar).</code></pre>
<p><img src="merge_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>Perhaps not surprisingly, when we combine the abundances of non-replicate samples from the same environment, the estimates of absolute richness increase for each environment. More to the example, however, the new merged plot is easier to read and interpret, which is one of the reasons one might use the <code>merge_samples</code> function.</p>
</div>
<div id="merge_taxa" class="section level2">
<h2>merge_taxa</h2>
<p>One of the sources of “noise” in a microbial census dataset is a fine-scaled definition for OTU that blurs a pattern that might be otherwise evident were we to consider a higher taxonomic rank. The best way to deal with this is to use the agglomeration functions <code>tip_glom</code> or <code>tax_glom</code> that merge similar OTUs based on a phylogenetic or taxonomic threshold, respectively. However, for either function to work, it must be capable of merging two or more OTUs that have been deemed “equivalent”.</p>
<p>The following is an example of <code>merge_taxa</code> in action, using a tree graphic to display the before and after of the merge, and also show how the merge affects not just the OTU table, but all data components with OTU indices.</p>
<p>We will use an example dataset that was previously imported into phyloseq with the variable name <code>closedps</code> and saved within an RData binary file, <code>&quot;example-data.RData&quot;</code>.</p>
<pre class="r"><code>load(&quot;example-data.RData&quot;)
plot_tree(closedps, color=&quot;Treatment&quot;, size=&quot;abundance&quot;, sizebase=2, label.tips=&quot;taxa_names&quot;)</code></pre>
<p><img src="merge_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>Now let’s use <code>merge_taxa</code> to merge the first 8 OTUS of <code>closedps</code> into one new OTU. By choosing <code>2</code> as the optional third argument to <code>merge_taxa</code>, we are combining the counts of these 8 OTUs into the index for the second OTU. The default is to use the first OTU of the second argument.</p>
<pre class="r"><code>x1 = merge_taxa(closedps, taxa_names(closedps)[3:27], 2)
plot_tree(x1, color=&quot;Treatment&quot;, size=&quot;abundance&quot;, sizebase=2, label.tips=&quot;taxa_names&quot;)</code></pre>
<p><img src="merge_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<hr />
</div>
<div id="merge_phyloseq" class="section level2">
<h2>merge_phyloseq</h2>
<p>As said earlier, <code>merge_phyloseq</code> is a convenience/support tool to help get your data into the right format. Here is an example in which we extract components from an example dataset, and then build them back up to the original form using <code>merge_phyloseq</code> along the way.</p>
<p>Let’s split apart the “Global Patterns” example dataset into some components</p>
<pre class="r"><code>data(GlobalPatterns)
tree = phy_tree(GlobalPatterns)
tax  = tax_table(GlobalPatterns)
otu  = otu_table(GlobalPatterns)
sam  = sample_data(GlobalPatterns)
otutax = phyloseq(otu, tax)
otutax</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
## tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]</code></pre>
<p>As you can see, our new <code>otutax</code> object has just the OTU table and taxonomy table. Now let’s use <code>merge_phyloseq</code> to build up the original <code>GlobalPatterns</code> object, and compare to make sure they are identical. Note how the arguments to <code>merge_phyloseq</code> are a mixture of multi-component (<code>otutax</code>) and single component objects.</p>
<pre class="r"><code>GP2 = merge_phyloseq(otutax, sam, tree)
identical(GP2, GlobalPatterns)</code></pre>
<pre><code>## [1] TRUE</code></pre>
<div id="more-complicated-merges" class="section level3">
<h3>More complicated merges</h3>
<p>The <code>merge_phyloseq</code> function will also work even with more than one multiple-component object.</p>
<pre class="r"><code>otusamtree = phyloseq(otu, sam, tree)
GP3 = merge_phyloseq(otusamtree, otutax)
GP3</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
## sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
## tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
## phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</code></pre>
<pre class="r"><code>identical(GP3, GlobalPatterns)</code></pre>
<pre><code>## [1] FALSE</code></pre>
<p>So this merge appears to have worked. The new object, <code>GP3</code>, looks similar to <code>GlobalPatterns</code> but is not identical. Why? Well, the assumption by <code>merge_phyloseq</code> is that you are atttempting to merge <strong>separate</strong> sources of abundance data, and so any portion of the OTU tables in the two phyloseq objects that have the same OTU indices are <strong>summed</strong> together, just like with <code>merge_taxa</code> earlier.</p>
<p>This example scenario was illustrative, but hopefully rare in practice. Nevertheless, just in case, an easy fix would be to extract the unique component of <code>otutax</code> and provide it to <code>merge_phyloseq</code>, instead of the entire phyloseq object. This amounts to a small intuitive modification to the previous <code>merge_phyloseq</code> command:</p>
<pre class="r"><code>GP4 = merge_phyloseq(otusamtree, tax_table(otutax))
GP4</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
## sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
## tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
## phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</code></pre>
<pre class="r"><code>identical(GP4, GlobalPatterns)</code></pre>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
