`r opts_chunk$set(cache=TRUE, fig.width=8, warning=FALSE)`
<link href="http://kevinburke.bitbucket.org/markdowncss/markdown.css" rel="stylesheet"></link>

# plot_tree function -
# Powerful tree graphics with ggplot2

This page contains many already-constructed examples of trees created by the `plot_tree` function in the phyloseq package, which in turn is uses the powerful graphics framework package called [ggplot2](http://docs.ggplot2.org/current/).

## The esophagus dataset.
A simple dataset containing tree and OTU-table components.

The esophagus is a small and relatively simple dataset by moderns standards. It only contains 3 samples, no sample-data, and a modest quantity of total sequencing per sample that is a relic of an earlier time when resources for this sort of investigation were sparse and sequencing was expensive. Nevertheless, it makes for a nice sized dataset to start displaying trees. (For more details about the dataset and its origin, try entering `?esophagus` into the command-line once you have loaded the phyloseq package)

Load the package and dataset

```{r, message=FALSE}
library("phyloseq")
data("esophagus")
```

The default tree without any additional parameters is plot with black points that indicate in how many different samples each OTU was found. In this case, the term "OTU" is used quite loosely (it is a loose term, after all) to mean entries in the taxononmic data you are plotting; and in the specific case of trees, it means the tips, even if you have already agglomerated the data such that each tip is equivalent to a rank of class or phylum. 

```{r}
plot_tree(esophagus, title="Default tree.")
```

If for some reason you just want a completely unadorned tree, the `"treeonly"` method can be selected. This tends to plot much faster than the annotated tree, and is still a ggplot2 object that you *might* be able to add further layers to manually.

```{r}
plot_tree(esophagus, "treeonly", title="method = \"treeonly\"")
```

Now let's shade tips according to the sample in which a particular OTU was observed.

```{r}
plot_tree(esophagus, color="samples")
```

We can also scale the size of tips according to abundance; usually related to number of sequencing reads, but depends on what you have done with the data prior to this step.

```{r}
plot_tree(esophagus, size="abundance")
```

Both graphical features included at once.

```{r}
plot_tree(esophagus, size="abundance", color="samples")
```

There is some overlap of the tip points. Let's adjust the base spacing to spread them out a little bit.

```{r}
plot_tree(esophagus, size="abundance", color="samples", base.spacing=0.03)
```

Good, now what if we wanted to also display the specific numeric value of OTU abundances that occurred more than 3 times in a given sample? For that, `plot_tree` includes the `min.abundance` parameter, set to `Inf` by default to prevent any point labels from being written.

```{r}
plot_tree(esophagus, size="abundance", color="samples", base.spacing=0.03, min.abundance=3)
```

# Examples with the Global Patterns dataset

Load the dataset (quotes are optional, did you know that?)

```{r}
data(GlobalPatterns)
```

Subset Global Patterns dataset to just the observed Archaea.

```{r}
gpa <- subset_species(GlobalPatterns, Kingdom=="Archaea")
```

The number of different Archaeal species from this dataset is small enough for a decent tree plot.

```{r}
nspecies(gpa)
```

That is to say, it is reasonable to consider displaying the phylogenetic tree directly for `gpa`. Too many OTUs means a tree that is pointless to attempt to display in its entirety in one graphic of a standard size and font. So the whole `GlobalPatterns` dataset probably a bad idea.

```{r}
nspecies(GlobalPatterns)
```

Some patterns are immediately discernable with minimal parameter choices:

```{r}
plot_tree(gpa, color="SampleType")
```


```{r}
plot_tree(gpa, color="Phylum")
```


```{r}
plot_tree(gpa, color="SampleType", shape="Phylum")
```


```{r}
plot_tree(gpa, color="Phylum", label.tips="Genus")
```

However, the text-label size scales with number of species, and with common graphics-divice sizes/resolutions, these ~200 taxa still make for a somewhat crowded graphic. 

Let's instead subset further to just the Crenarchaeota

```{r}
gpac <- subset_species(gpa, Phylum=="Crenarchaeota")
plot_tree(gpac, color="SampleType", shape="Genus")
```


```{r}
plot_tree(gpac, color="SampleType", label.tips="Genus")
```

Let's add some abundance information. Notice that the default spacing gets a little crowded when we map species-abundance to point-size:

```{r}
plot_tree(gpac, color="SampleType", shape="Genus", size="abundance")
```

So let's spread it out a little bit with the base.spacing parameter.

```{r}
plot_tree(gpac, color="SampleType", shape="Genus", size="abundance", base.spacing=0.05)
```

## Chlamydiae-only tree

```{r, fig.width=10}
GP.chl <- subset_species(GlobalPatterns, Phylum=="Chlamydiae")
plot_tree(GP.chl, color="SampleType", shape="Family", label.tips="Genus", size="abundance")
```			
			

## Other key graphics functions in the phyloseq package:

### [plot_ordination](http://joey711.github.com/phyloseq/plot_ordination-examples)

### [plot_tree](http://joey711.github.com/phyloseq/plot_tree-examples)

### [plot_network](http://joey711.github.com/phyloseq/plot_network-examples)

### [plot_richness](http://joey711.github.com/phyloseq/plot_richness-examples)

### [plot_bar](http://joey711.github.com/phyloseq/plot_bar-examples)
