<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Heatmap Plots</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<link href="site_libs/ionicons-2.0.1/css/ionicons.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}

.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">phyloseq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="install.html">
    <span class="ion ion-android-archive"></span>
     
    Installation
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-newspaper-o"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0061217">phyloseq in Bioconductor</a>
    </li>
    <li>
      <a href="http://bioinformatics.oxfordjournals.org/content/early/2014/09/26/bioinformatics.btu616">Shiny-phyloseq</a>
    </li>
    <li>
      <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003531">Rarefying Inadmissible</a>
    </li>
    <li>
      <a href="http://f1000research.com/articles/5-1492/v1">Full Stack Workflow</a>
    </li>
    <li>
      <a href="https://www.nature.com/articles/nmeth.3869">DADA2</a>
    </li>
    <li>
      <a href="https://www.nature.com/articles/ismej2017119">Amplicon Sequence Variants &gt; OTUs</a>
    </li>
  </ul>
</li>
<li>
  <a href="courses.html">
    <span class="fa fa-graduation-cap"></span>
     
    Courses
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-university"></span>
     
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Basics</li>
    <li>
      <a href="import-data.html">Data Import</a>
    </li>
    <li>
      <a href="Example-Data.html">Example Data</a>
    </li>
    <li>
      <a href="merge.html">Merge</a>
    </li>
    <li>
      <a href="preprocess.html">Preprocess</a>
    </li>
    <li>
      <a href="distance.html">Distances</a>
    </li>
    <li>
      <a href="gap-statistic.html">Gap Statistic</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Workflows</li>
    <li>
      <a href="http://benjjneb.github.io/dada2/tutorial.html">DADA2</a>
    </li>
    <li>
      <a href="http://f1000research.com/articles/5-1492/v1">Full Stack</a>
    </li>
    <li>
      <a href="http://joey711.github.io/phyloseq-demo/HMP_import_example.html">HMP-1</a>
    </li>
    <li>
      <a href="http://joey711.github.io/phyloseq-demo/Restroom-Biogeography.html">Restroom Biogeography</a>
    </li>
    <li>
      <a href="http://joey711.github.io/phyloseq-demo/unifrac.html">UniFrac</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Graphics</li>
    <li>
      <a href="plot_ordination-examples.html">Plot Ordination</a>
    </li>
    <li>
      <a href="plot_richness-examples.html">Plot Alpha Diversity</a>
    </li>
    <li>
      <a href="plot_tree-examples.html">Plot Tree</a>
    </li>
    <li>
      <a href="plot_bar-examples.html">
        <span class="fa fa-bar-chart"></span>
         
        Plot Bars
      </a>
    </li>
    <li>
      <a href="plot_heatmap-examples.html">
        <span class="fa fa-qrcode"></span>
         
        Plot Heatmap
      </a>
    </li>
    <li>
      <a href="plot_network-examples.html">
        <span class="ion ion-network"></span>
         
        Plot Network
      </a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Differential Abundance</li>
    <li>
      <a href="http://joey711.github.io/phyloseq-extensions/DESeq2.html">DESeq2 (Recommended)</a>
    </li>
    <li>
      <a href="http://joey711.github.io/phyloseq-extensions/edgeR.html">edgeR</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/joey711/phyloseq/issues">
    <span class="fa fa-medkit"></span>
     
    Issues
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/phyloseq/inst/doc/phyloseq-FAQ.html">
    <span class="fa fa-question"></span>
     
    FAQ
  </a>
</li>
<li>
  <a href="https://github.com/joey711/phyloseq">
    <span class="fa fa-github-alt"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-twitter"></span>
     
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://twitter.com/joey711">
        <span class="fa fa-twitter"></span>
         
      </a>
    </li>
    <li>
      <a href="https://www.linkedin.com/in/joey711">
        <span class="fa fa-linkedin-square"></span>
         
      </a>
    </li>
    <li>
      <a href="http://stackoverflow.com/users/935950/paul-mcmurdie">
        <span class="fa fa-stack-overflow"></span>
         
      </a>
    </li>
    <li>
      <a href="https://scholar.google.com/citations?user=nNQ1LtMAAAAJ">
        <span class="ai ai-google-scholar-square"></span>
         
      </a>
    </li>
    <li>
      <a href="https://www.researchgate.net/profile/Paul_Mcmurdie">
        <span class="ai ai-researchgate-square"></span>
         
      </a>
    </li>
    <li>
      <a href="https://github.com/joey711/phyloseq">
        <span class="fa fa-github-alt"></span>
         
      </a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Heatmap Plots</h1>
<h4 class="date"><em>Mon Mar 12 15:07:17 2018</em></h4>

</div>


<p>The following demonstrates some uses of the <code>plot_heatmap</code> function in <a href="http://joey711.github.com/phyloseq/">the phyloseq package</a> for R and Bioconductor.</p>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<div id="an-ecologically-organized-heatmap" class="section level2">
<h2>An ecologically-organized heatmap</h2>
<p>In a <a href="http://www.biomedcentral.com/1471-2105/11/45">2010 article in BMC Genomics</a>, Rajaram and Oono describe an approach to creating a heatmap using ordination methods (namely, NMDS and PCA) to organize the rows and columns instead of (hierarchical) cluster analysis. In many cases the ordination-based ordering does a much better job than h-clustering at providing an order of elements that is easily interpretable. The authors provided an immediately useful example of their approach as <a href="http://cran.r-project.org/web/packages/NeatMap/index.html">the NeatMap package for R</a>. The NeatMap package can be used directly on the abundance table (<code>&quot;otu_table&quot;</code>-class) of phylogenetic-sequencing data, but the NMDS or PCA ordination options that it supports are not based on ecological distances. To fill this void, and because phyloseq already provides support for a large number of <a href="http://joey711.github.io/phyloseq/distance">ecological distances</a> and <a href="http://joey711.github.io/phyloseq/ordinate">ordination methods</a>, phyloseq now includes the <code>plot_heatmap()</code> function: an ecology-oriented variant of the NeatMap approach to organizing a heatmap and build it using ggplot2 graphics tools. The <a href="http://joey711.github.io/phyloseq/distance">distance</a> and <a href="http://joey711.github.io/phyloseq/ordinate">method</a> arguments are the same as for the <a href="http://joey711.github.io/phyloseq/plot_ordination">plot_ordination</a> function, and support large number of distances and ordination methods, respectively, with a strong leaning toward ecology. This function also provides the options to re-label the OTU and sample axis-ticks with a taxonomic name and/or sample variable, respectively, in the hope that this might hasten your interpretation of the patterns (See the documentation for the <code>sample.label</code> and <code>taxa.label</code> arguments, and the examples below). Note that this function makes no attempt to overlay dendrograms from hierarchical clustering next to the axes, as hierarchical clustering is not used to organize these plots. Also note that each re-ordered axis repeats at the edge, and so apparent clusters at the far right/left or top/bottom of the heat-map may actually be the same. For now, the placement of this edge can be considered arbitrary, so beware of this artifact of the graphic and visually check if there are two “mergeable” clusters at the edges of a particular axis. If you benefit from this phyloseq-specific implementation of <a href="http://cran.r-project.org/web/packages/NeatMap/index.html">the NeatMap approach</a>, please cite <a href="http://www.biomedcentral.com/1471-2105/11/45">the NeatMap article</a>, as well as phyloseq.</p>
</div>
<div id="heatmap-colors-dont-have-to-be-so-hot" class="section level2">
<h2>Heatmap colors don’t have to be so hot</h2>
<p>Traditionally heatmaps have been used to emphasize data that is above or below a threshold as “hot” or “cold” colors, respectively. In the case of OTU-abundance data, however, it seems the most common need is to see the relative patterns of high-abundance OTUs against a background of taxa that are mostly low-abundance or absent in a sparse matrix. Furthermore, there is usually not an obvious or intrinsically meaningful abundance value to use as a suitable threshold for the traditional “cold/hot” display. For these reasons, the default color scheme in <code>plot_heatmap</code> maps a very dark blue color to the lowest abundance values, up to a very light blue for the highest abundance values. This default continuous color palette is a coherent, blue-oriented color scheme in which the eye should be drawn to the lighter shades. The dark blue for the lowest abundance values is not very much lighter than black, and true black is used to represent missing or zero abundance values.</p>
<p>If, for whatever reason, you need to change this default color scheme, it is possible through the <code>low</code>, <code>high</code>, and <code>na.value</code> arguments. Several examples are provided below. The character-string values supplied to these arguments need to be the names of R colors. There are over 600 English color names that are understood by R (try <code>colors()</code> at the R terminal), as well as other finely-resolved color gradient nomenclatures. The examples below use a 6-digit hexadecimal color representation, and a nice [table summary of these colors is available at the R Cookbook](<a href="http://wiki.stdout.org/rcookbook/Graphs/Colors%20(ggplot2\)/">http://wiki.stdout.org/rcookbook/Graphs/Colors%20(ggplot2\)/</a>).</p>
<p>For further details, not that the <code>high</code>, <code>low</code>, and <code>na.value</code> parameters are passed along to <a href="http://had.co.nz/ggplot2/scale_gradient.html">ggplot2’s scale_gradient function</a>, which does an excellent job selecting suitable colors of your gradient, provided that you select colors that make sense to have at two ends of a gradient.</p>
<p>I also got some useful ideas and suggestions at the following <a href="http://learnr.wordpress.com/2010/01/26/ggplot2-quick-heatmap-plotting/">WordPress page regarding the construction of heatmaps using ggplot2’s</a> <code>geom_tile</code>.</p>
</div>
<div id="color-scaling" class="section level2">
<h2>Color scaling</h2>
<p>In addition to the color palette that defines the poles, color in the heatmap is also characterized by the numerical transformation from observed value to color – called <em>color scaling</em>.</p>
<p>By default, the <code>plot_heatmap</code> color scale is a log transformation with base 4, using <code>log_trans(4)</code> from <a href="http://cran.r-project.org/web/packages/scales/index.html">the scales package</a>. This is an arbitrary choice that you might need to adjust based on your needs and data. If specifying an alternative transformation object to the <code>trans</code> argument, you probably need to load the scales package first. Since <a href="http://cran.r-project.org/web/packages/scales/index.html">scales</a> is a required package for <code>phyloseq</code>, you should already have it installed if you are at this point. Any transformation object that is valid for the scales package should work here, but the relative contrast and the way it represents your data could change dramatically based on this choice, so make this selection carefully; or better yet, try several different transformations if you think data is being “left in the background” or too much information is being “pushed to the foreground”, for example.</p>
</div>
</div>
<div id="examples" class="section level1">
<h1>Examples</h1>
<div id="load-packages-data" class="section level2">
<h2>Load Packages, Data</h2>
<pre class="r"><code>library(&quot;phyloseq&quot;); packageVersion(&quot;phyloseq&quot;)</code></pre>
<pre><code>## [1] &#39;1.22.3&#39;</code></pre>
<pre class="r"><code>library(&quot;ggplot2&quot;); packageVersion(&quot;ggplot2&quot;)</code></pre>
<pre><code>## [1] &#39;2.2.1&#39;</code></pre>
<p>ggplot2 package theme set. See <a href="http://docs.ggplot2.org/current/">the ggplot2 online documentation</a> for further help.</p>
<pre class="r"><code>theme_set(theme_bw())</code></pre>
</div>
<div id="plot-a-300-taxa-dataset" class="section level2">
<h2>Plot a 300-taxa dataset</h2>
<p>The following two lines subset the dataset to just the top 300 most abundant Bacteria taxa across all samples (in this case, with no prior preprocessing. Not recommended, but quick).</p>
<pre class="r"><code>data(&quot;GlobalPatterns&quot;)
gpt &lt;- subset_taxa(GlobalPatterns, Kingdom==&quot;Bacteria&quot;)
gpt &lt;- prune_taxa(names(sort(taxa_sums(gpt),TRUE)[1:300]), gpt)
plot_heatmap(gpt, sample.label=&quot;SampleType&quot;)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
</div>
<div id="subset-a-smaller-dataset-based-on-an-archaeal-phylum" class="section level2">
<h2>Subset a smaller dataset based on an Archaeal phylum</h2>
<p>Subset the dataset to something manageable that can be reasonably represented in one plot. In the following examples, the Crenarchaeota phylum.</p>
<pre class="r"><code>gpac &lt;- subset_taxa(GlobalPatterns, Phylum==&quot;Crenarchaeota&quot;)</code></pre>
</div>
<div id="default-plot_heatmap-settings" class="section level2">
<h2>Default <code>plot_heatmap</code> settings</h2>
<p>Now let’s see how our <code>plot_heatmap</code> function works with all default settings.</p>
<pre class="r"><code>plot_heatmap(gpac)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/unnamed-chunk-4-1.png" width="384" /></p>
</div>
<div id="re-label-by-a-sample-variable-and-taxonomic-family" class="section level2">
<h2>Re-label by a sample variable and taxonomic family</h2>
<p>Here is an example re-labelling based on the “SampleType” sample variable and the taxonomic rank of “Family”.</p>
<pre class="r"><code>(p &lt;- plot_heatmap(gpac, &quot;NMDS&quot;, &quot;bray&quot;, &quot;SampleType&quot;, &quot;Family&quot;))</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
</div>
<div id="re-label-axis-titles" class="section level2">
<h2>Re-label axis titles</h2>
<p>What if you wanted to change the axis labels, but not the labels on individual features?</p>
<pre class="r"><code>p$scales$scales[[1]]$name &lt;- &quot;My X-Axis&quot;
p$scales$scales[[2]]$name &lt;- &quot;My Y-Axis&quot;
print(p)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/GPheatmap-rename-axes-1.png" width="672" /></p>
</div>
<div id="now-repeat-the-plot-but-change-the-color-scheme." class="section level2">
<h2>Now repeat the plot, but change the color scheme.</h2>
<p>Changing the color scheme might be worthwhile, depending on the graphics device or paper on which you want to display the heatmap.</p>
<pre class="r"><code>plot_heatmap(gpac, &quot;NMDS&quot;, &quot;bray&quot;, &quot;SampleType&quot;, &quot;Family&quot;, low=&quot;#000033&quot;, high=&quot;#CCFF66&quot;)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>Here is a dark-blue to red scheme.</p>
<pre class="r"><code>plot_heatmap(gpac, &quot;NMDS&quot;, &quot;bray&quot;, &quot;SampleType&quot;, &quot;Family&quot;, low=&quot;#000033&quot;, high=&quot;#FF3300&quot;)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>A very dark-blue to very light-blue scheme</p>
<pre class="r"><code>plot_heatmap(gpac, &quot;NMDS&quot;, &quot;bray&quot;, &quot;SampleType&quot;, &quot;Family&quot;, low=&quot;#000033&quot;, high=&quot;#66CCFF&quot;)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<p>Here is a “dark on light” color scheme. Note that we change the background value (the value of the NA and 0 elements)</p>
<pre class="r"><code>plot_heatmap(gpac, &quot;NMDS&quot;, &quot;bray&quot;, &quot;SampleType&quot;, &quot;Family&quot;, low=&quot;#66CCFF&quot;, high=&quot;#000033&quot;, na.value=&quot;white&quot;)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<p>This is a similar color scheme as the previous, but the “near zero” color is closer to a cream color, and the colors themselves are closer to blue-grey. This is better overall contrast than a lot of schemes, but may not be as exciting.</p>
<pre class="r"><code>plot_heatmap(gpac, &quot;NMDS&quot;, &quot;bray&quot;, &quot;SampleType&quot;, &quot;Family&quot;, low=&quot;#FFFFCC&quot;, high=&quot;#000033&quot;, na.value=&quot;white&quot;)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
</div>
<div id="now-try-different-ordination-methods-distances" class="section level2">
<h2>Now try different ordination methods, distances</h2>
<p>Now try the default color scheme, but using different ecological distances/ordinations. For example, NMDS ordination on the jaccard distance.</p>
<pre class="r"><code>plot_heatmap(gpac, &quot;NMDS&quot;, &quot;jaccard&quot;)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<p>Detrended correspondence analysis.</p>
<pre class="r"><code>plot_heatmap(gpac, &quot;DCA&quot;, &quot;none&quot;, &quot;SampleType&quot;, &quot;Family&quot;)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<p>Unconstrained redundancy analysis (Principle Components Analysis, PCA)</p>
<pre class="r"><code>plot_heatmap(gpac, &quot;RDA&quot;, &quot;none&quot;, &quot;SampleType&quot;, &quot;Family&quot;)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<p>PCoA/MDS ordination on the (default) bray-curtis distance.</p>
<pre class="r"><code>plot_heatmap(gpac, &quot;PCoA&quot;, &quot;bray&quot;, &quot;SampleType&quot;, &quot;Family&quot;)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/gpac-MDS-1.png" width="672" /></p>
<p>MDS/PCoA ordination on the Unweighted-UniFrac distance.</p>
<pre class="r"><code>plot_heatmap(gpac, &quot;PCoA&quot;, &quot;unifrac&quot;, &quot;SampleType&quot;, &quot;Family&quot;)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<p>Now try weighted-UniFrac distance and MDS/PCoA ordination.</p>
<pre class="r"><code>plot_heatmap(gpac, &quot;MDS&quot;, &quot;unifrac&quot;, &quot;SampleType&quot;, &quot;Family&quot;, weighted=TRUE)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<p>Here is how you might create a heatmap using base-R graphics and the more common (but problematic) hierarchical clustering organization, in case you want to compare with <code>plot_heatmap</code>, for example.</p>
<pre class="r"><code>heatmap(otu_table(gpac))</code></pre>
<p><img src="plot_heatmap-examples_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
