library(phyloseq)
library(ggplot2)


## The esophagus dataset.
The esophagus is a small and relatively simple dataset by moderns standards. It only contains 3 samples, no sample-data, and a modest quantity of total sequencing per sample that is a relic of an earlier time when resources for this sort of investigation were sparse and  sequencing was expensive. Nevertheless, it makes for a nice sized dataset to start displaying trees.

Load the package and dataset

```{r, message=FALSE}
library("phyloseq")
data("esophagus")
```

Here is the basic tree plot without any additional parameters.
```{r}
plot_tree(esophagus)
```

Now let's shade tips according to the sample in which a particular OTU was observed.

```{r}
plot_tree(esophagus, color="samples")
```

We can also scale the size of tips according to abundance; usually related to number of sequencing reads, but depends on what you have done with the data prior to this step.

```{r}
plot_tree(esophagus, size="abundance")
```

Both graphical features included at once.

```{r}
plot_tree(esophagus, size="abundance", color="samples")
```

Adjust the base spacing.

```{r}
plot_tree(esophagus, size="abundance", color="samples", base.spacing=0.03)
```

# Examples with the Global Patterns dataset

Load the dataset (quotes are optional, did you know that?)

```{r}
data(GlobalPatterns)
```

Subset Global Patterns dataset to just the observed Archaea.

```{r}
gpa <- subset_species(GlobalPatterns, Kingdom=="Archaea")
```

The number of different Archaeal species from this dataset is small enough for a decent tree plot.

```{r}
nspecies(gpa)
```

That is to say, it is reasonable to consider displaying the phylogenetic tree directly for `gpa`. Too many OTUs means a tree that is pointless to attempt to display in its entirety in one graphic of a standard size and font. So the whole `GlobalPatterns` dataset probably a bad idea.

```{r}
nspecies(GlobalPatterns)
```

Some patterns are immediately discernable with minimal parameter choices:

```{r}
plot_tree(gpa, color="SampleType")
```


```{r}
plot_tree(gpa, color="Phylum")
```


```{r}
plot_tree(gpa, color="SampleType", shape="Phylum")
```


```{r}
plot_tree(gpa, color="Phylum", label.tips="Genus")
```

However, the text-label size scales with number of species, and with common graphics-divice sizes/resolutions, these ~200 taxa still make for a somewhat crowded graphic. 

Let's instead subset further ot just the Crenarchaeota

```{r}
gpac <- subset_species(gpa, Phylum=="Crenarchaeota")
plot_tree(gpac, color="SampleType", shape="Genus")
```


```{r}
plot_tree(gpac, color="SampleType", label.tips="Genus")
```

Let's add some abundance information. Notice that the default spacing gets a little crowded when we map species-abundance to point-size:

```{r}
plot_tree(gpac, color="SampleType", shape="Genus", size="abundance")
```

So let's spread it out a little bit with the base.spacing parameter.

```{r}
plot_tree(gpac, color="SampleType", shape="Genus", size="abundance", base.spacing=0.05)
```

## Chlamydiae-only tree

```{r}
GP.chl <- subset_species(GlobalPatterns, Phylum=="Chlamydiae")
plot_tree(GP.chl, color="SampleType", shape="Family", label.tips="Genus", size="abundance")
```			
			

## Other key graphics functions in the phyloseq package:

### [plot_ordination](http://joey711.github.com/phyloseq/plot_ordination-examples)

### [plot_tree](http://joey711.github.com/phyloseq/plot_tree-examples)

### [plot_network](http://joey711.github.com/phyloseq/plot_network-examples)

### [plot_richness](http://joey711.github.com/phyloseq/plot_richness-examples)

### [plot_bar](http://joey711.github.com/phyloseq/plot_bar-examples)
